flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesRunningJobsRegistry.java

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * {@link RunningJobsRegistry} implementation for Kubernetes. All the running jobs will be stored in
+ * Dispatcher-leader ConfigMap. The key is the job id with prefix
+ * {@link org.apache.flink.kubernetes.utils.Constants#RUNNING_JOBS_REGISTRY_KEY_PREFIX},
+ * and value is job status.
+ */
+public class KubernetesRunningJobsRegistry implements RunningJobsRegistry {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesRunningJobsRegistry.class);
+
+	private final FlinkKubeClient kubeClient;
+
+	private final String configMapName;
+
+	private final String lockIdentity;
+
+	public KubernetesRunningJobsRegistry(FlinkKubeClient kubeClient, String configMapName, String lockIdentity) {
+		this.kubeClient = checkNotNull(kubeClient);
+		this.configMapName = checkNotNull(configMapName);
+		this.lockIdentity = checkNotNull(lockIdentity);
+	}
+
+	@Override
+	public void setJobRunning(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		writeJobStatusToConfigMap(jobID, JobSchedulingStatus.RUNNING);
+	}
+
+	@Override
+	public void setJobFinished(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		writeJobStatusToConfigMap(jobID, JobSchedulingStatus.DONE);
+	}
+
+	@Override
+	public JobSchedulingStatus getJobSchedulingStatus(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		return kubeClient.getConfigMap(configMapName)
+			.map(configMap -> {
+				final String key = getKeyForJobId(jobID);
+				final String status = configMap.getData().get(key);
+				if (!StringUtils.isNullOrWhitespaceOnly(status)) {
+					return JobSchedulingStatus.valueOf(status);
+				} else {
+					LOG.warn("ConfigMap {} contained corrupted data. Ignoring the key {}.", configMapName, key);
+				}
+				return JobSchedulingStatus.PENDING;
+			})
+			.orElseThrow(() -> new IOException(
+				new KubernetesException("ConfigMap " + configMapName + " does not exist.")));
+	}
+
+	@Override
+	public void clearJob(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		try {
+			kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				configMap -> {
+					if (KubernetesLeaderElector.hasLeadership(configMap, lockIdentity)) {
+						configMap.getData().remove(getKeyForJobId(jobID));
+						return Optional.of(configMap);
+					}
+					return Optional.empty();
+				}
+			).get();
+		} catch (Exception e) {
+			throw new IOException("Failed to clear job state in ConfigMap " + configMapName + " for job " + jobID, e);
+		}
+	}
+
+	private void writeJobStatusToConfigMap(JobID jobID, JobSchedulingStatus status) throws IOException {
+		LOG.debug("Setting scheduling state for job {} to {}.", jobID, status);
+		final String key = getKeyForJobId(jobID);
+		try {
+			kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				configMap -> {
+					if (KubernetesLeaderElector.hasLeadership(configMap, lockIdentity)) {
+						configMap.getData().put(key, status.name());
+						return Optional.of(configMap);
```

Comment:
Might be better to check the original value and try to avoid the update if the value is not changed.